mirrord/layer/src/socket.rs

```diff
@@ -499,6 +537,40 @@ impl OutgoingSelector {
             .collect()),
         }
     }
+
+    /// Helper function that looks into the [`DNS_CACHE`] for `address`, so we can retrieve the
+    /// hostname and resolve it locally (when applicable).
+    ///
+    /// We only get here when the [`OutgoingSelector::Remote`] matched nothing, thus
+    /// `selected_address` comes in as `None`, or when the [`OutgoingSelector::Local`] matched on
+    /// something, so we have `Some(selected_address)`.
+    ///
+    /// Returns 1 of 3 possibilities:
+    ///
+    /// 1. `address` is in [`DNS_CACHE`]: resolves the hostname locally, then return it as
+    /// [`ConnectionThrough::Local`];
+    /// 2. `address` is **NOT** in [`DNS_CACHE`]: check `REMOTE` and return the valid local address;
+    #[tracing::instrument(level = "trace", ret)]
+    fn get_address_to_connect<const REMOTE: bool>(
+        address: SocketAddr,
+        selected_address: Option<SocketAddr>,
+    ) -> Detour<ConnectionThrough> {
+        if let Some((cached_hostname, port)) = DNS_CACHE
+            .get(&SocketAddr::from((address.ip(), 0)))
+            .map(|addr| (addr.value().clone(), address.port()))
+        {
+            let _guard = DetourGuard::new();
+            let locally_resolved = format!("{cached_hostname}:{port}")
+                .to_socket_addrs()?
+                .find(SocketAddr::is_ipv4)?;
+
+            Detour::Success(ConnectionThrough::Local(locally_resolved))
```

Comment:
Yeah, it became messier the more I tried it out locally.

Think if we get the PR itself addressed (pun intended), then someone else can jump in just to review the tests, wdyt?

Reply:
My (first) comment is about the code potentially being incorrect, so it's not just a matter of having tests. Tests could justify a less thorough review, but if we don't have tests, we should at least have an explanation of why what I wrote is not a problem and the code is correct.

Reply:
1. user sets `remote = ["service"]`:

- `connect(service)` matches, then we `connect(address)`, where `address` is whatever `service` resolved to normally (depends on dns enabled);
- `connect(service)` no matches, then we `connect(local_address)`;

2. user sets `local = ["service"]`:

- `connect(service)` matches, then we `connect(local_address)`;
- `connect(service)` no matches, then we `connect(address)`, where `address` is whatever `service` resolved to normally (depends on dns enabled); <- this is the case you're mentioning right?

Reply:
Yes that's the case, only that the code doesn't look like it depends on the DNS setting, it looks like in that case we re-resolve the host locally, and then connect locally:
https://github.com/metalbear-co/mirrord/blob/70968410bef86410e11f9ab74c10a7f0ec2d277a/mirrord/layer/src/socket.rs#L561-L566

Reply:
But don't we only get there if there is a match + filter is set to local?

```rust
 if list
  .iter()
  .filter(skip_unresolved)
  .chain(hosts)
  .filter(filter_protocol)
  .any(any_address)
  {
    Self::get_local_address_to_connect(address)?
```

Otherwise we are going to connect to whatever `address` was passed down from `connect`, which means that it's the `address` resolved through the normal DNS procedure (and depends on dns feature).

Reply:
https://github.com/metalbear-co/mirrord/pull/1648/files#diff-bf9c5f9340f82cf6357f9ea7b2d2a735776603ed1ead3daa2cd95db99c024650R439-R449

Reply:
Not sure how to put the nice code snippet like you did here :cry:

Reply:
Ok, I think you're right, it's fine.

Reply:
> Not sure how to put the nice code snippet like you did here 😢

You need to use a permalink